public/friendlycode/js/fc/filesystem-sync.js

```diff
@@ -103,22 +114,124 @@ define(["jquery"], function($) {
-        }
+        trigger(fsync._callbacks.beforeEach, arguments);
 
         Array.prototype.unshift.call(arguments, url, csrfToken, fs);
         handler.apply(fsync, arguments);
       };
     }
     Bramble.once("ready", function(bramble) {
-      bramble.on("fileChange", configHandler(pushFileChange, persistanceUrls.createOrUpdate));
-      bramble.on("fileDelete", configHandler(pushFileDelete, persistanceUrls.del));
-      bramble.on("fileRename", configHandler(pushFileRename, persistanceUrls.rename));
+      fsync.bramble = bramble;
+      fsync.handlers = {
+        change: configHandler(pushFileChange, persistanceUrls.createOrUpdate),
+        del: configHandler(pushFileDelete, persistanceUrls.del),
+        rename: configHandler(pushFileRename, persistanceUrls.rename)
+      };
+
+      bramble.on("fileChange", fsync.handlers.change);
+      bramble.on("fileDelete", fsync.handlers.del);
+      bramble.on("fileRename", fsync.handlers.rename);
     });
 
     return fsync;
   };
 
+  FSync.prototype.saveAndSyncAll = function(callback) {
+    var fsync = this;
+    var bramble = fsync.bramble;
+    var afterEach = fsync._callbacks.afterEach;
+    var filesSaved = [];
+
+    function resetAfterEach(fn) {
+      var index = afterEach.indexOf(fn);
+      if(index !== -1) {
+        afterEach.splice(index, 1);
+      }
+    }
+
+    function maybeFinish(err, path) {
+      function finish() {
+        resetAfterEach(maybeFinish);
+        bramble.on("fileChange", fsync.handlers.change);
+        callback.apply(null, arguments);
+      }
+
+      if(err) {
+        finish(err, path);
+        return;
+      }
+
+      if(path) {
+        filesSaved.splice(filesSaved.indexOf(path), 1);
+      }
+
+      if(filesSaved.length === 0) {
+        finish();
+      }
+    }
+
+    function syncAll() {
+      if(filesSaved.length === 0) {
+        maybeFinish();
+        return;
+      }
+
+      fsync.afterEach(maybeFinish);
+      filesSaved.forEach(function(path) {
+        fsync.handlers.change(path);
+      });
+    }
+
+    function cacheFilesSaved(path) {
+      filesSaved.push(path);
+    }
+
+    function maybeContinue(err) {
+      if(err) {
+        resetAfterEach(maybeContinue);
+        callback(err);
+        return;
+      }
+
+      if(fsync.queueLength > 0) {
+        return;
+      }
+
+      resetAfterEach(maybeContinue);
+      bramble.off("fileChange", fsync.handlers.change);
+      bramble.on("fileChange", cacheFilesSaved);
+
+      bramble.saveAll(function() {
+        bramble.off("fileChange", cacheFilesSaved);
+        syncAll();
+      });
+    }
+
+    // If there are syncs queued up, wait on them
+    if(fsync.queueLength > 0) {
+      fsync.afterEach(maybeContinue);
+    } else {
+      maybeContinue();
+    }
+  };
+
+  // Add a callback to execute before every sync
+  // Each callback receives a `path` as an argument
+  FSync.prototype.beforeEach = function(callback) {
+    if(typeof callback !== "function") {
```

Comment:
I don't think this check is necessary; same is true below for `afterEach`.  I might change the names, though: `FSync.prototype.add{before|after}EachCallback`